src/nvim/api/vim.c

```diff
@@ -2389,3 +2389,140 @@ Dictionary nvim_complete_set(Integer index, Dict(complete_set) *opts, Arena *are
+  if (wp->w_lines_valid == 0) {
+    *flush = true;
+  }
+
+  // Mark for redraw in case flush will happen, otherwise redraw now.
+  if (*flush && (opts->statusline || opts->winbar)) {
+    wp->w_redr_status = true;
+  } else if (opts->statusline || opts->winbar) {
+    win_check_ns_hl(wp);
+    if (opts->winbar) {
+      win_redr_winbar(wp);
+    }
+    if (opts->statusline) {
+      win_redr_status(wp);
+    }
+    win_check_ns_hl(NULL);
+  }
+}
+
+/// Instruct Nvim to redraw various components.
+///
+/// @param opts  Optional parameters.
+///               - win: Target a specific |window-ID| as described below.
+///               - buf: Target a specific buffer number as described below.
+///               - flush: Update the screen with pending updates.
+///               - clear: Forcefully redraw `win`, all windows containing
+///                 `buf`, or all windows.
+///               - range: Redraw a range in `buf`, the buffer in `win` or
+///                 the current buffer. Useful for decoration providers.
+///                 Expects a tuple `[first, last]` with the first and last
+///                 line number of the range.
+///               - cursor: Immediately update cursor position on the screen
+///                 in `win` or the current window.
+///               - statuscolumn: Redraw the 'statuscolumn' in `buf`, `win` or
+///                 all windows.
+///               - statusline: Redraw the 'statusline' in `buf`, `win` or all
+///                 windows.
+///               - winbar: Redraw the 'winbar' in `buf`, `win` or all windows.
+///               - tabline: Redraw the 'tabline'.
+void nvim_redraw(Dict(redraw) *opts, Error *err)
+  FUNC_API_SINCE(12)
+{
+  win_T *win = NULL;
+  buf_T *buf = NULL;
+
+  if (HAS_KEY(opts, redraw, win)) {
+    win = find_window_by_handle(opts->win, err);
+    if (ERROR_SET(err)) {
+      return;
+    }
+  }
+
+  if (HAS_KEY(opts, redraw, buf)) {
+    VALIDATE(win == NULL, "%s", "cannot use both 'buf' and 'win'", {
+      return;
+    });
+    buf = find_buffer_by_handle(opts->buf, err);
+    if (ERROR_SET(err)) {
+      return;
+    }
+  }
+
+  if (opts->clear) {
+    if (win != NULL) {
+      redraw_later(win, UPD_NOT_VALID);
+    } else if (buf != NULL) {
+      redraw_buf_later(buf, UPD_NOT_VALID);
+    } else {
+      redraw_all_later(UPD_NOT_VALID);
+    }
+  }
+
+  if (HAS_KEY(opts, redraw, range)) {
+    VALIDATE(kv_size(opts->range) == 2
+             && kv_A(opts->range, 0).type == kObjectTypeInteger
+             && kv_A(opts->range, 1).type == kObjectTypeInteger,
+             "%s", "Invalid 'range': Expected 2-tuple of Integers", {
+      return;
+    });
+    linenr_T first = (linenr_T)kv_A(opts->range, 0).data.integer;
+    linenr_T last = (linenr_T)kv_A(opts->range, 1).data.integer;
+    if (last < 0) {
+      last = buf->b_ml.ml_line_count;
+    }
+    redraw_buf_range_later(win ? win->w_buffer : (buf ? buf : curbuf), first + 1, last);
```

Comment:
I think it should be reconsidered what `first` and `last` actually indicate, and how they are documented. It seems that in `nvim__buf_redraw_range` `first` was a 0-indexed row number, whereas `last` was a 1-indexed line number (why?). I'll go ahead and document it and update the internal calls like so:

```
- range: Redraw a range in `buf`, the buffer in `win` or
  the current buffer. Useful for decoration providers.
  Expects a tuple `[first, last]` with the first and last
  line number of the range, or [1, -1] to indicate the entire
  buffer.
```

Reply:
Ah yeah that makes more sense probably, so then they were both 0-indexed. But I think 1-indexed end-inclusive line numbers makes more sense for this API?

Reply:
I mean in a way it doesn't because the extmarks that necessitate this call also were placed 0-indexed (end inclusive). I just mean in general when talking about line numbers 1-indexed feels more natural but if this should be 0-indexed as well it just needs to be documented.

Reply:
It's also not entirely clear to me why this API is needed in the first place so that probably doesn't help. Don't the `nvim_buf_set_extmark()` calls already set `w_redraw_top/mod` through `decor_redraw()`? If this API is graduated the reasoning behind that should also probably be documented better (here)?

Reply:
I think all internal `nvim__buf_redraw_range(buf, 0, -1)` calls could/should be replaced by something like:

```diff
@@ -2425,8 +2425,9 @@ static void redraw_status(win_T *wp, Dict(redraw) *opts, bool *flush)
+///               - valid: When present and `true`, redraw the changed part of `win`,
+///                 all windows containing `buf`, or all windows. When `false`, redraw
+///                 the entire window.
 ///               - range: Redraw a range in `buf`, the buffer in `win` or the
 ///                 current buffer. Useful for decoration providers. Expects a
 ///                 tuple `[first, last]` with the first and last line number of
@@ -2462,13 +2463,14 @@ void nvim_redraw(Dict(redraw) *opts, Error *err)
     }
   }

-  if (opts->clear) {
+  if (HAS_KEY(opts, redraw, valid)) {
+    int type = opts->valid ? UPD_VALID : UPD_NOT_VALID;
     if (win != NULL) {
-      redraw_later(win, UPD_NOT_VALID);
+      redraw_later(win, type);
     } else if (buf != NULL) {
-      redraw_buf_later(buf, UPD_NOT_VALID);
+      redraw_buf_later(buf, type);
     } else {
-      redraw_all_later(UPD_NOT_VALID);
+      redraw_all_later(type);
     }
   }
```

Then, if placing ephemeral marks would call `decor_redraw()`, I don't think this `range` API would be needed.

Reply:
It also seems like the `nvim__buf_redraw_range(buf, 0, -1)` calls were "abusing" the private API. I will try to replace `clear` with `valid` in a way that I think should cover that use case.